probe-rs/src/architecture/arm/core/m4.rs

```diff
@@ -636,28 +658,32 @@ impl<'probe> CoreInterface for M4<'probe> {
         Architecture::Arm
     }
 
-    /// Read the hardware breakpoints from FpComp registers, and adds them to the Result Vector. A value of 0 in any position of the Vector indicates that the position is unset/available.
-    fn get_hw_breakpoints(&mut self) -> Result<Vec<u32>, Error> {
+    /// See docs on the [`CoreInterface::get_hw_breakpoints`] trait
+    fn get_hw_breakpoints(&mut self) -> Result<Vec<Option<u32>>, Error> {
         let mut breakpoints = vec![];
         let num_hw_breakpoints = self.get_available_breakpoint_units()? as usize;
         { 0..num_hw_breakpoints }.try_for_each(|bp_unit_index| {
             let raw_val = self.memory.read_word_32(FpCtrl::ADDRESS)?;
             let ctrl_reg = FpCtrl::from(raw_val);
-            let breakpoint: u32;    // The breakpoint address after it has been adjusted for FpRev
-            let register_value: u32;// The raw breakpoint address as read from memory
+            // FpRev1 and FpRev2 needs different decoding of the register value, but the location where we read from is the same ...
+            let reg_addr = FpRev1CompX::ADDRESS + (bp_unit_index * size_of::<u32>()) as u32;
+            // The raw breakpoint address as read from memory
+            let register_value = self.memory.read_word_32(reg_addr)?;
+            // The breakpoint address after it has been adjusted for FpRev 1 or 2
+            let breakpoint:u32;
             if ctrl_reg.rev() == 0 {
-                let reg_addr = FpRev1CompX::ADDRESS + (bp_unit_index * size_of::<u32>()) as u32;
-                register_value = self.memory.read_word_32(reg_addr)?;
-                breakpoint = register_value & 0x1f_ff_ff_fc;
+                breakpoint = FpRev1CompX::get_breakpoint_comparator(register_value)?;
             } else if ctrl_reg.rev() == 1 {
-                let reg_addr = FpRev2CompX::ADDRESS + (bp_unit_index * size_of::<u32>()) as u32;
-                register_value = self.memory.read_word_32(reg_addr)?;
-                breakpoint = register_value & 0xff_ff_ff_fe;
+                breakpoint = FpRev2CompX::from(register_value).bpaddr() << 1;
             } else {
                 log::warn!("This chip uses FPBU revision {}, which is not yet supported. HW breakpoints are not available.", ctrl_reg.rev());
                 return Err(Error::Other(anyhow!("This chip uses FPBU revision {}, which is not yet supported. HW breakpoints are not available.", ctrl_reg.rev())));
             }
-            breakpoints.push(breakpoint);
+            if breakpoint > 0 {
```

Comment:
I think we should check the enable bit, instead of checking if the value is 0. Someone could set a breakpoint at address 0.

Reply:
Done :)